microprofile/microprofile-fault-tolerance-2.1/src/main/java/org/wildfly/swarm/ts/microprofile/fault/tolerance/v21/retry/RetryService.java

```diff
@@ -0,0 +1,165 @@
+package org.wildfly.swarm.ts.microprofile.fault.tolerance.v21.retry;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import javax.enterprise.context.RequestScoped;
+
+import org.eclipse.microprofile.faulttolerance.Asynchronous;
+import org.eclipse.microprofile.faulttolerance.Fallback;
+import org.eclipse.microprofile.faulttolerance.Retry;
+import org.wildfly.swarm.ts.microprofile.fault.tolerance.v21.MyConnection;
+import org.wildfly.swarm.ts.microprofile.fault.tolerance.v21.MyException;
+
+@Asynchronous // all is async with fallback method
+@RequestScoped
+public class RetryService {
+
+    private AtomicInteger counter = new AtomicInteger(1);
+
+    // First fail, second pass - there will be 1 retry before success
+    @Retry(retryOn = Throwable.class)
+    public CompletionStage<MyConnection> throwable_retryOnT() throws Throwable {
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new Throwable("Simulated Throwable");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from throwable_retryOnT, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    // First fail, second pass - there will be 1 retry before success (Throwable is parent of Exception)
+    @Retry(retryOn = Throwable.class)
+    @Fallback(fallbackMethod = "processFallback")
+    public CompletionStage<MyConnection> myException_retryOnT() throws MyException {
+        int invocationNumber = counter.getAndIncrement();
+        if (invocationNumber % 2 == 1) {
+            throw new MyException("Simulated custom Exception");
+        }
+
+        return CompletableFuture.completedFuture(new MyConnection() {
+            @Override
+            public String getData() {
+                return "Hello from myException_retryOnT, counter: " + invocationNumber;
+            }
+        });
+    }
+
+    // First call goes through and is recognized by fallback
+    @Retry(abortOn = Throwable.class)
+    @Fallback(fallbackMethod = "processFallback")
+    public CompletionStage<MyConnection> throwable_abortOnT() throws Throwable {
```

Comment:
So this should fail once and then immediately go to fallback, yet you check that fallback is `Hello from fallback, counter: 2`. That sounds weird to me.

Reply:
I didn't check if other methods are also weird like this. Could you please doublecheck?

Reply:
I suggest to start counter from 0 and instead of int invocationNumber = counter.getAndIncrement() use counter.incrementAndGet() .. it should solve all weirdness

Reply:
I used the counter badly. Now it should be ok with Petrs solution. Thank you both!